APCAppCore/APCAppCore/Library/DataArchiverAndUploader/APCDataArchive.m

```diff
@@ -0,0 +1,267 @@
+// 
+ 
+#import "APCDataArchive.h"
+#import "zipzap.h"
+#import <objc/runtime.h>
+#import "APCUtilities.h"
+#import "ORKAnswerFormat+Helper.h"
+#import "APCLog.h"
+#import "NSDate+Helper.h"
+#import "APCJSONSerializer.h"
+#import "NSError+APCAdditions.h"
+
+static NSString * kFileInfoNameKey                  = @"filename";
+static NSString * kUnencryptedArchiveFilename       = @"unencrypted.zip";
+static NSString * kFileInfoTimeStampKey             = @"timestamp";
+static NSString * kFileInfoContentTypeKey           = @"contentType";
+static NSString * kTaskRunKey                       = @"taskRun";
+static NSString * kFilesKey                         = @"files";
+static NSString * kAppNameKey                       = @"appName";
+static NSString * kAppVersionKey                    = @"appVersion";
+static NSString * kPhoneInfoKey                     = @"phoneInfo";
+static NSString * kItemKey                          = @"item";
+
+@interface APCDataArchive ()
+
+@property (nonatomic, strong) NSString *reference;
+@property (nonatomic, strong) ZZArchive *zipArchive;
+@property (nonatomic, strong) NSMutableArray *zipEntries;
+@property (nonatomic, strong) NSMutableArray *filesList;
+@property (nonatomic, strong) NSMutableDictionary *infoDict;
+
+@end
+
+@implementation APCDataArchive
+
+//designated initializer
+- (id)initWithReference: (NSString *)reference
+{
+    self = [super init];
+    if (self) {
+        _reference = reference;
+        [self createArchive];
+    }
+    
+    return self;
+}
+
+//create a new zip archive at the reference path
+- (void)createArchive
+{
+    
+    NSURL *zipArchiveURL = [NSURL fileURLWithPath:[[self workingDirectoryPath] stringByAppendingPathComponent:kUnencryptedArchiveFilename]];
+    _unencryptedURL = zipArchiveURL;
+
+    _zipEntries = [NSMutableArray array];
+    _filesList = [NSMutableArray array];
+    _infoDict = [NSMutableDictionary dictionary];
+    NSError * error;
+    
+    _zipArchive = [[ZZArchive alloc] initWithURL:zipArchiveURL
+                                             options:@{ZZOpenOptionsCreateIfMissingKey : @YES}
+                                               error:&error];
+    if (!_zipArchive) {
+        APCLogError2(error);
+    }
+}
+
+//A sandbox in the temporary directory for this archive to be cleaned up on completion.
+- (NSString *)workingDirectoryPath
+{
+    
+    NSString *workingDirectoryPath = [NSTemporaryDirectory() stringByAppendingPathComponent:self.reference];
+    
+    if (![[NSFileManager defaultManager] fileExistsAtPath:workingDirectoryPath]) {
+        NSError * fileError;
+        BOOL created = [[NSFileManager defaultManager] createDirectoryAtPath:workingDirectoryPath withIntermediateDirectories:YES attributes:@{ NSFileProtectionKey : NSFileProtectionComplete } error:&fileError];
+        if (!created) {
+            APCLogError2 (fileError);
+        }
+    }
+    
+    return workingDirectoryPath;
+}
```

Comment:
Fixed but not showing here - only in full changes. Set workingDirectoryPath to nil if !created.